File: day21/missingNumber.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <random>
#include <ctime>

int missingNumber(const std::vector<int>& nums) {
    int n = nums.size();
    int expected_sum = n * (n + 1) / 2;
    int actual_sum = std::accumulate(nums.begin(), nums.end(), 0);
    return expected_sum - actual_sum;
}

int main() {
    std::mt19937 rng(static_cast<unsigned int>(std::time(nullptr)));
    std::uniform_int_distribution<int> sizeDist(10, 30);

    int n = sizeDist(rng);
    std::vector<int> full(n + 1);
    std::iota(full.begin(), full.end(), 0); // Fill with 0 to n

    std::uniform_int_distribution<int> removeDist(0, n);
    int missing = full[removeDist(rng)];
    full.erase(std::remove(full.begin(), full.end(), missing), full.end());

    std::shuffle(full.begin(), full.end(), rng);

    std::cout << "Input: ";
    for (int num : full) std::cout << num << " ";
    std::cout << "\nExpected missing number: " << missing << "\n";
    std::cout << "Computed missing number: " << missingNumber(full) << "\n";

    return 0;
}

File: day21/MissingNumber.java
import java.util.*;

public class FindMissing {

    public static int missingNumber(int[] nums) {
        int n = nums.length;
        int expectedSum = n * (n + 1) / 2;
        int actualSum = 0;
        for (int num : nums) actualSum += num;
        return expectedSum - actualSum;
    }

    public static void main(String[] args) {
        Random rand = new Random();
        int n = rand.nextInt(21) + 10; // n between 10 and 30
        List<Integer> full = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            full.add(i);
        }

        int missing = full.remove(rand.nextInt(full.size())); // remove one random number
        Collections.shuffle(full);
        int[] nums = full.stream().mapToInt(Integer::intValue).toArray();

        System.out.println("Input: " + Arrays.toString(nums));
        System.out.println("Expected missing number: " + missing);
        System.out.println("Computed missing number: " + missingNumber(nums));
    }
}
